streaming/in.go

```diff
@@ -296,20 +295,22 @@ func (c *crypt4GHInternalReader) read(p []byte) (n int, err error) {
 				}
 				haveRead++
 			}
-		} else {
-			// Read larger chunk from buffer. As precaution, limit to what we
-			// should be able to read only, as that is the bit we've checked
-			// if the discard list imposes any holes in
-
+		case canRead > 0:
 			r, err := c.buffer.Read(p[haveRead : haveRead+canRead])
 			haveRead += r
-			c.streamPos += int64(r)
+			if err != nil {
+				return haveRead, err
+			}
 
-			// Not sure why we'd get an error here, but forward
-			// if we see that
+			c.streamPos += int64(r)
+		case canRead == 0:
+			r, err := c.buffer.Read(p[haveRead:])
+			haveRead += r
 			if err != nil {
 				return haveRead, err
 			}
+
+			c.streamPos += int64(r)
```

Comment:
Needs to be above error return, methinks.

Reply:
See above.